### PR TITLE
Android: Joystick improvements

### PR DESCRIPTION
**Pull request by garbear**
This improves joystick logging on Android and fixes several bugs:

* Fix ignoring some buttons on the PS4 controller
* Fix off-by-1 controller number
* Fix dropping input while debugging

The third bug was due to a race condition that only appeared when running an unoptimized debug build. Input polling was too slow, so button presses were being dropped. Now all presses are preserved.

Joysticks now appear in the log like this:

```
Debug Print: CPeripheralBusAndroid: Device added:
Debug Print:   Name: "Wireless Controller"
Debug Print:     ID: 11
Debug Print:     Controller number: 1
Debug Print:     Descriptor: "3088fa86f07d61..."
Debug Print:     Product ID: 09CC
Debug Print:     Vendor ID: 054C
Debug Print:     Has microphone: false
Debug Print:     Is virtual: false
Debug Print:     Source flags: 0x01000511
Debug Print:     Has source: SOURCE_GAMEPAD (0x00000401)
Debug Print:     Has source: SOURCE_JOYSTICK (0x01000010)
Debug Print:     Has source: SOURCE_KEYBOARD (0x00000101)
Debug Print:     Has key: AKEYCODE_BUTTON_A (96)
Debug Print:     Has key: AKEYCODE_BUTTON_B (97)
Debug Print:     Has key: AKEYCODE_BUTTON_C (98)
Debug Print:     Has key: AKEYCODE_BUTTON_X (99)
Debug Print:     Has key: AKEYCODE_BUTTON_Y (100)
Debug Print:     Has key: AKEYCODE_BUTTON_Z (101)
Debug Print:     Has key: AKEYCODE_BUTTON_L1 (102)
Debug Print:     Has key: AKEYCODE_BUTTON_R1 (103)
Debug Print:     Has key: AKEYCODE_BUTTON_L2 (104)
Debug Print:     Has key: AKEYCODE_BUTTON_R2 (105)
Debug Print:     Has key: AKEYCODE_BUTTON_THUMBL (106)
Debug Print:     Has key: AKEYCODE_BUTTON_START (108)
Debug Print:     Has key: AKEYCODE_BUTTON_SELECT (109)
Debug Print:     Has key: AKEYCODE_BUTTON_MODE (110)
Debug Print:     Has axis: AMOTION_EVENT_AXIS_X (0)
Debug Print:       Endpoints: [-1.000000, 1.000000]
Debug Print:       Center: 0.117647
Debug Print:       Fuzz: 0.000000
Debug Print:     Has axis: AMOTION_EVENT_AXIS_Y (1)
Debug Print:       Endpoints: [-1.000000, 1.000000]
Debug Print:       Center: 0.117647
Debug Print:       Fuzz: 0.000000
Debug Print:     Has axis: AMOTION_EVENT_AXIS_Z (11)
Debug Print:       Endpoints: [-1.000000, 1.000000]
Debug Print:       Center: 0.117647
Debug Print:       Fuzz: 0.000000
Debug Print:     Has axis: AMOTION_EVENT_AXIS_RX (12)
Debug Print:       Endpoints: [-1.000000, 1.000000]
Debug Print:       Center: 0.117647
Debug Print:       Fuzz: 0.000000
Debug Print:     Has axis: AMOTION_EVENT_AXIS_RY (13)
Debug Print:       Endpoints: [-1.000000, 1.000000]
Debug Print:       Center: 0.117647
Debug Print:       Fuzz: 0.000000
Debug Print:     Has axis: AMOTION_EVENT_AXIS_RZ (14)
Debug Print:       Endpoints: [-1.000000, 1.000000]
Debug Print:       Center: 0.117647
Debug Print:       Fuzz: 0.000000
Debug Print:     Has axis: AMOTION_EVENT_AXIS_HAT_X (15)
Debug Print:       Endpoints: [-1.000000, 1.000000]
Debug Print:       Center: 0.000000
Debug Print:       Fuzz: 0.000000
Debug Print:     Has axis: AMOTION_EVENT_AXIS_HAT_Y (16)
Debug Print:       Endpoints: [-1.000000, 1.000000]
Debug Print:       Center: 0.000000
Debug Print:       Fuzz: 0.000000
Debug Print: CPeripheralBusAndroid: input device with ID 11 added
```

Requires https://github.com/xbmc/libandroidjni/pull/14

## Motivation and Context
I plugged in a PS4 controller and hacked away until it worked better.

## How Has This Been Tested?
Tested on a Samsung Galaxy S8 with a PS4 controller in debug (unoptimized) build.

Test builds available at: https://github.com/garbear/xbmc/releases

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -275,6 +275,7 @@ cmake_install.cmake
 /tools/depends/**/config.log
 /tools/depends/**/config.status
 /tools/depends/native/*/*native/
+/tools/depends/target/*/aarch64-linux-*/
```
Comment (Rechi) [inline .gitignore]: add this to tools/depends/.gitignore

File: tools/depends/target/libandroidjni/Makefile [hunk 2]
```diff
@@ -3,7 +3,7 @@ DEPS= ../../Makefile.include Makefile
 
 # lib name, version
 LIBNAME=libandroidjni
-VERSION=bfe01d03cfcef423ffbf3cd9aafef4a5a3539201
+VERSION=54f4894d929e95f25371f0c95ff100e132150a38
```
Comment (Rechi) [inline tools/depends/target/libandroidjni/Makefile]: has to be adjusted to the merge ref once merged

File: xbmc/platform/android/peripherals/PeripheralBusAndroid.cpp [hunk 3]
```diff
@@ -329,28 +344,111 @@ bool CPeripheralBusAndroid::GetDeviceId(const std::string& deviceLocation, int&
...
+  else
+    CLog::Log(LOGDEBUG, "    Descriptor: \"%s\"", descriptor.c_str());
+  CLog::Log(LOGDEBUG, "    Product ID: %04X", device.getProductId());
+  CLog::Log(LOGDEBUG, "    Vendor ID: %04X", device.getVendorId());
+  CLog::Log(LOGDEBUG, "    Has microphone: %s", device.hasMicrophone() ? "true" : "false");
+  CLog::Log(LOGDEBUG, "    Is virtual: %s", device.isVirtual() ? "true" : "false");
+
+  // Log device sources
+  CLog::Log(LOGDEBUG, "    Source flags: 0x%08x", device.getSources());
+  for (auto source : GetInputSources())
```
Comment (Rechi) [inline xbmc/platform/android/peripherals/PeripheralBusAndroid.cpp]: `const auto& source`

**Comment by garbear** [comment]
Test builds uploaded: https://github.com/garbear/xbmc/releases

Got a report of a crash at https://forum.kodi.tv/showthread.php?tid=173361&pid=2759782#pid2759782

```
V/CrashDescriptorUtil(  518): signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x19998c4
V/CrashDescriptorUtil(  518): Abort message: 'art/runtime/check_jni.cc:72] JNI DETECTED ERROR IN APPLICATION: JNI CallBooleanMethodV called with pending exception 'java.lang.NoSuchMethodError' thrown in unknown throw location'
V/CrashDescriptorUtil(  518):
V/CrashDescriptorUtil(  518): backtrace:
...
V/CrashDescriptorUtil(  518):     #13 pc 00000000019e0da8  /data/app/org.xbmc.kodi-2/lib/arm64/libkodi.so (_ZNK19CJNIViewInputDevice13hasMicrophoneEv+28)
```

It seems CJNIViewInputDevice::hasMicrophone() is missing, but it has been in libandroidjni from before https://github.com/xbmc/libandroidjni/pull/14...

File: .gitignore (hunk 1 above)
Comment (garbear) [reply .gitignore]: updated

**Comment by garbear** [comment]
merged https://github.com/xbmc/libandroidjni/pull/14. jenkins build this please

File: xbmc/platform/android/peripherals/AndroidJoystickTranslator.cpp [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this Program; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
```
Comment (Rechi) [inline xbmc/platform/android/peripherals/AndroidJoystickTranslator.cpp]: wrong copyright notice

File: xbmc/platform/android/peripherals/AndroidJoystickTranslator.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this Program; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
```
Comment (Rechi) [inline xbmc/platform/android/peripherals/AndroidJoystickTranslator.h]: wrong copyright notice

Comment (garbear) [reply xbmc/platform/android/peripherals/AndroidJoystickTranslator.h]: fixed

**Comment by garbear** [comment]
jenkins build this please

**Comment by garbear** [comment]
Dropped 01e792353fbbefaf183fcf47eef6fbdade8307ac - This removed the android axis filtering and used Kodi's axis filtering, but Kodi's filtering wasn't able to handle triggers before they're pressed (resulting in erratic motion until both triggers were pressed and depressed)

File: xbmc/peripherals/devices/PeripheralJoystick.cpp [hunk 6]
```diff
@@ -70,10 +70,18 @@ bool CPeripheralJoystick::InitialiseFeature(const PeripheralFeature feature)
   {
     if (feature == FEATURE_JOYSTICK)
     {
-      if (m_bus->InitializeProperties(*this))
-        bSuccess = true;
+      // Ensure an add-on is present to translate input
+      if (!m_manager.GetAddonWithButtonMap(this))
+      {
+        CLog::Log(LOGERROR, "CPeripheralJoystick: No button mapping add-on for %", m_strLocation.c_str());
```
Comment (peak3d) [inline xbmc/peripherals/devices/PeripheralJoystick.cpp]: This or the next CLog are too early, CLog does not exist at the moment this Log is called -> segfault on shield during startup

Comment (garbear) [reply xbmc/peripherals/devices/PeripheralJoystick.cpp]: @FernetMenta is CLog a global we can kill?

**Comment by garbear** [comment]
@peak3d does reverting 3c7ebc5ec2a169be29d900f7e333b177df9561d2 fix the issue?

**Comment by peak3d** [comment]
@garbear the issue appeared during hacking with some other android components, its currently not reproducable on my systems - so hard to say.

But because the issue appeared once, it will most likely come back sooner or later

**Comment by garbear** [comment]
The peripheral callback should be registered before this log statement is ever hit. Maybe I'm wrong and you can track down a violation. Of course, there could be a memory leak in an unrelated part of the code. Often when debugging I hit these kind of snafus.

**Pull request merged**